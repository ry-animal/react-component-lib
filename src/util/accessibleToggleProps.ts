export function accessibleToggleProps(
  label: string,
  controls: string,
  expanded: boolean,
  popupType?: "menu"
) {
  return {
    "aria-controls": controls,
    "aria-expanded": expanded,
    "aria-haspopup": popupType || true,
    "aria-label": label,
    "aria-pressed": expanded,
    role: "button",
  };
}
