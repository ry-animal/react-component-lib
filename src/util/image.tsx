import { ISelectedFile } from "src/components/Forms/FileUpload/ISelectedFile";

export const makeDataUrl = (selectedFile: ISelectedFile) =>
  URL.createObjectURL(selectedFile.file);
