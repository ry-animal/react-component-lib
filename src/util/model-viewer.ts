/*
 * This file is a hacky way of allowing any other component to
 * use the <model-viewer /> tag just by doing:
 *   import "../../../util/model-viewer";
 *
 * The authors of the package do not expose it in a React-friendly
 * way, see: https://github.com/google/model-viewer/issues/1502
 *
 * The component documentation is at https://modelviewer.dev/docs/
 */

import "@google/model-viewer";

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace JSX {
    export interface IntrinsicElements {
      "model-viewer": {
        src?: string;
        alt?: string;
        poster?: string;
        autoplay?: boolean;
        exposure?: number;
        "camera-controls"?: boolean;
        "environment-image"?: string;
        "seamless-poster"?: string;
        loading?: "auto" | "lazy" | "eager";
        reveal?: "auto" | "interaction" | "manual";
      } & React.DetailedHTMLProps<
        React.HTMLAttributes<HTMLElement>,
        HTMLElement
      >;
    }
  }
}
