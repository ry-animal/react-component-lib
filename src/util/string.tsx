const DEFAULT_LOCALE = "en-US";

export const truncateEthAddress = (str: string): string => {
  if (str.length) {
    const numStartingCharacters = 6;
    const numTrailingCharacters = 4;
    return `${str.substring(0, numStartingCharacters)}...${str.substring(
      str.length - numTrailingCharacters,
      str.length
    )}`;
  }
  return "";
};

export const formatUSDPriceString = (value: number): string => {
  const minimumFractionDigits = 2;
  const maximumFractionDigits = 2;
  return `$${value.toLocaleString(DEFAULT_LOCALE, {
    minimumFractionDigits,
    maximumFractionDigits,
  })}`;
};

export const formatETHPriceString = (priceInEth: number): string => {
  const minimumFractionDigits = 0;
  const maximumFractionDigits = 8;
  const maxDigitsWithDecimal = 8;

  let formatted = `${priceInEth
    .toLocaleString(DEFAULT_LOCALE, {
      minimumFractionDigits,
      maximumFractionDigits,
    })
    .replace(/,/gu, "")
    .substring(0, maxDigitsWithDecimal)}`;

  if (formatted.endsWith(".")) {
    formatted = formatted.slice(0, -1);
  }
  return formatted;
};

const evaluateNumberLessThanOne = (num: number) => {
  if (num === 0) {
    return "0";
  }

  const minDecimalValue = 0.0001;
  if (num < minDecimalValue) {
    return minDecimalValue.toString();
  }
  // To round decimal to 4 digits
  const maxDigits = 10000;
  return `${Math.round(num * maxDigits) / maxDigits}`;
};

// Abbreviate Number https://stackoverflow.com/a/9462382

export const abbreviateNumber = (num: number): string => {
  const lookup = [
    { value: 1, symbol: "" },
    { value: 1e3, symbol: "k" },
    { value: 1e6, symbol: "M" },
    { value: 1e9, symbol: "B" },
    { value: 1e12, symbol: "T" },
    { value: 1e15, symbol: "P" },
    { value: 1e18, symbol: "E" },
  ];

  const rx = /\.0+$|(?<id>\.[0-9]*[1-9])0+$/u;
  const digits = 1;

  const item = lookup
    .slice()
    .reverse()
    .find((x) => num >= x.value);
  return item
    ? (num / item.value).toFixed(digits).replace(rx, "$1") + item.symbol
    : evaluateNumberLessThanOne(num);
};

export const getDomainFromRegex = (url: string) => {
  const matches = url.match(/:\/\/(?<id>.[^/]+)/u);
  // Extract domain (will be null if no match is found)
  return matches && matches[1];
};
