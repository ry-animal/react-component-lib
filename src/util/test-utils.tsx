import { render, RenderOptions } from "@testing-library/react";
import "jest-styled-components";
import * as React from "react";
import { BrowserRouter } from "react-router-dom";
import { ThemeProvider } from "styled-components";
import { LIGHT_THEME } from "../theme";

interface ProvidersProps {
  children: React.ReactNode;
}

function Providers({ children }: ProvidersProps) {
  return (
    <BrowserRouter>
      <ThemeProvider theme={LIGHT_THEME}>{children}</ThemeProvider>
    </BrowserRouter>
  );
}

const customRender = (
  ui: JSX.Element,
  options?:
    | RenderOptions<
        typeof import("@testing-library/dom/types/queries"),
        HTMLElement
      >
    | undefined
) => render(ui, { wrapper: Providers, ...options });

// Duplicated instead of import "../../__mocks__/matchMediaMock"; because jest has trouble resolving imports
Object.defineProperty(window, "matchMedia", {
  writable: true,
  value: jest.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: jest.fn(),
    removeListener: jest.fn(),
    addEventListener: jest.fn(),
    removeEventListener: jest.fn(),
    dispatchEvent: jest.fn(),
  })),
});

// Re-export everything
export * from "@testing-library/react";
// Override render method
export { customRender as render };
