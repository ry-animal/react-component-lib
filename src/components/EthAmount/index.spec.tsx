import { EthAmount } from ".";
import { render } from "../../util/test-utils";

describe("<EthAmount />", () => {
  describe("rendering", () => {
    test("Default rendering matches snapshot", () => {
      expect(
        render(<EthAmount amount={0.0001} />).asFragment()
      ).toMatchSnapshot();
    });

    test("Bold rendering matches snapshot", () => {
      expect(
        render(<EthAmount amount={0.0001} bold />).asFragment()
      ).toMatchSnapshot();
    });
  });
});
