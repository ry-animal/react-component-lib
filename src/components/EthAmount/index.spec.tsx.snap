// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EthAmount /> rendering Bold rendering matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Poppins',sans-serif;
  font-weight: 600;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #151515;
}

<span
    class="c0"
  >
    <svg
      fill="none"
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        clip-path="url(#clip0_14895_64936)"
      >
        <path
          d="M12 16.4939V20.5L16.8571 13.5929L12 16.4939Z"
          data-testid="icon-eth"
          fill="#151515"
          fill-opacity="0.8"
        />
        <path
          d="M12 10.4252V15.5671L16.8571 12.6584L12 10.4252Z"
          fill="#151515"
          fill-opacity="0.3"
        />
        <path
          d="M12 4.5V10.4252L16.8571 12.6584L12 4.5Z"
          fill="#151515"
        />
        <path
          d="M11.9997 16.4939V20.5L7.14258 13.5929L11.9997 16.4939Z"
          fill="#151515"
          opacity="0.3"
        />
        <path
          d="M11.9997 10.4252V15.5671L7.14258 12.6584L11.9997 10.4252Z"
          fill="#151515"
          opacity="0.1"
        />
        <path
          d="M11.9997 4.5V10.4252L7.14258 12.6584L11.9997 4.5Z"
          fill="#151515"
          opacity="0.6"
        />
      </g>
      <defs>
        <clippath
          id="clip0_14895_64936"
        >
          <rect
            fill="#151515"
            height="16"
            transform="translate(7.14258 4.5)"
            width="9.71429"
          />
        </clippath>
      </defs>
    </svg>
    0.0001
  </span>
</DocumentFragment>
`;

exports[`<EthAmount /> rendering Default rendering matches snapshot 1`] = `
<DocumentFragment>
  .c0 {
  font-family: 'Poppins',sans-serif;
  font-weight: 400;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  color: #151515;
}

<span
    class="c0"
  >
    <svg
      fill="none"
      height="20"
      viewBox="0 0 24 24"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <g
        clip-path="url(#clip0_14895_64936)"
      >
        <path
          d="M12 16.4939V20.5L16.8571 13.5929L12 16.4939Z"
          data-testid="icon-eth"
          fill="#151515"
          fill-opacity="0.8"
        />
        <path
          d="M12 10.4252V15.5671L16.8571 12.6584L12 10.4252Z"
          fill="#151515"
          fill-opacity="0.3"
        />
        <path
          d="M12 4.5V10.4252L16.8571 12.6584L12 4.5Z"
          fill="#151515"
        />
        <path
          d="M11.9997 16.4939V20.5L7.14258 13.5929L11.9997 16.4939Z"
          fill="#151515"
          opacity="0.3"
        />
        <path
          d="M11.9997 10.4252V15.5671L7.14258 12.6584L11.9997 10.4252Z"
          fill="#151515"
          opacity="0.1"
        />
        <path
          d="M11.9997 4.5V10.4252L7.14258 12.6584L11.9997 4.5Z"
          fill="#151515"
          opacity="0.6"
        />
      </g>
      <defs>
        <clippath
          id="clip0_14895_64936"
        >
          <rect
            fill="#151515"
            height="16"
            transform="translate(7.14258 4.5)"
            width="9.71429"
          />
        </clippath>
      </defs>
    </svg>
    0.0001
  </span>
</DocumentFragment>
`;
